### PR TITLE
[CONTRIB] Add PopenWorker process recycling

### PR DESCRIPTION
**Pull request by petersalas**
Background processes (e.g. builds during autotuning) can sometimes accumulate state or leak memory. This adds the ability for `PopenWorker` processes to optionally be recycled after a specified number of uses.

### CODE REVIEW

File: python/tvm/contrib/popen_pool.py
Comment (AndrewZhaoLuo) [inline python/tvm/contrib/popen_pool.py:226]: we can move this block below lines 248-250 and remove self._proc check. It will also immediately free resources after last remaining use instead of needing additional send call.

File: python/tvm/contrib/popen_pool.py
Comment (AndrewZhaoLuo) [reply python/tvm/contrib/popen_pool.py:97]: Hmm, should there be more warning about what "recycling" means? 

Like what if someone has a very stateful function and moving things to init state is undesirable.

**Review by AndrewZhaoLuo — COMMENTED** [review]
Haha, I think this is a parsimonious (but good) workaround to the issue. It would be cool if we could open an issue with an example B) if you have time.

**Comment by AndrewZhaoLuo** [comment]
Wonder what others think though? @shingjan ?

File: python/tvm/contrib/popen_pool.py
Comment (petersalas) [reply python/tvm/contrib/popen_pool.py:226]: We can't quite do that because we have to wait for the `recv` call (at the start of `send` is "ok" only by relying on the contract above):

```
        The caller must call recv before calling the next send in
        order to make sure the timeout and child process exit
        won't affect the later requests.
```

But, I can move this to the end of `recv` if you'd like? I'm indifferent -- I put it in `send` to try consolidate (most) of the logic in one function but I certainly don't feel strongly about it.

Comment (AndrewZhaoLuo) [reply python/tvm/contrib/popen_pool.py:226]: Ah I see, nah let's keep it consolidated

File: python/tvm/contrib/popen_pool.py
Comment (petersalas) [reply python/tvm/contrib/popen_pool.py:144]: Without it the behavior is (very) non-deterministic (TIL if the process hasn't exited when its `__del__` runs then it gets added to a list(!) that gets polled when future `Popen`s are created) -- but indeed it should eventually get reaped. (Granted there's no guarantee that it exits 1s after SIGKILL either but I was reluctant to block on it forever...)

Having spent time debugging accumulating zombie processes in the past I'm generally loathe to leave them...but I'm good with removing it. LMK what you think

Comment (AndrewZhaoLuo) [reply python/tvm/contrib/popen_pool.py:144]: Hmm, so freeing the pid entry is non-deterministic in the other way. I see. I think being this explicit has definite advantages then.

**Comment by zxybazh** [comment]
Very interesting, we observe the same behaviour with PopenWorker during tuning and our workaround is also very similar. Would be happy to learn about how you guys figured out the problem and do you have any trace on the root cause or potential fix. A minimum reproducible example would also be appreciated!

**Review by shingjan — COMMENTED** [review]
Thanks for sending in this fix! Memory leak around `PopenPool` has been an issue identified in #11074. Does this fix imply that `PopenExecutor` underlying process will be a zombie process if used too many times? If so what is the best heuristic number for max_number_use?

**Comment by junrushao1994** [comment]
That's exactly one issue that we recently identified. Thank you so much for this fix! Also CC @tqchen as the original author of PopenWorker

**Comment by AndrewZhaoLuo** [comment]
Do we have an existing issue for the memory leak stuff? I think based on some internal data we suspect it's in `tvm.lower` somewhere.

**Comment by zxybazh** [comment]
Just created one #11096.

**Review by AndrewZhaoLuo — APPROVED** [review]
Any objections? Otherwise will merge by eod

**Comment by junrushao1994** [comment]
LGTM too! BTW, if `maximum_uses` is set, while the process is run async (e.g. MetaSchedule RPCRunner), we might want some logic to make sure synchronization happens before recycling the child processes, which is beyond the scope of this PR

**Comment by junrushao1994** [comment]
Thanks @petersalas for bringing this workaround, and @AndrewZhaoLuo @shingjan @zxybazh for review and discussion :-))

**Pull request merged**
